### PR TITLE
xdsclient: fix LRS stream leaks when errors are encountered

### PR DESCRIPTION
**Pull request by dfawley**
RELEASE NOTES:
* xdsclient: fix goroutine leaks when load reporting is enabled

### CODE REVIEW

File: xds/internal/xdsclient/controller/version/v2/loadreport.go [hunk 1]
```diff
@@ -149,5 +152,17 @@ func (v2c *client) SendLoadStatsRequest(s grpc.ClientStream, loads []*load.Data)
 
 	req := &lrspb.LoadStatsRequest{ClusterStats: clusterStats}
 	v2c.logger.Infof("lrs: sending LRS loads: %+v", pretty.ToJSON(req))
-	return stream.Send(req)
+	if err := stream.Send(req); err != nil {
+		return getStreamError(stream)
```
Comment (easwars) [inline xds/internal/xdsclient/controller/version/v2/loadreport.go:156]: s/getStreamError/getStreamRecvError/?

File: xds/internal/xdsclient/controller/version/v2/loadreport.go [hunk 2]
```diff
@@ -149,5 +152,17 @@ func (v2c *client) SendLoadStatsRequest(s grpc.ClientStream, loads []*load.Data)
...
+	}
+	return nil
+}
+
+func getStreamError(stream lrsStream) error {
+	for {
+		_, err := stream.Recv()
+		if err != nil {
```
Comment (easwars) [inline xds/internal/xdsclient/controller/version/v2/loadreport.go:164]: Combine these two lines.

File: xds/internal/xdsclient/controller/version/v2/loadreport.go (hunk 1 above)
Comment (easwars) [reply xds/internal/xdsclient/controller/version/v2/loadreport.go:156]: What is trying to achieve?

File: xds/internal/xdsclient/controller/version/v2/loadreport.go [hunk 3]
```diff
@@ -149,5 +152,17 @@ func (v2c *client) SendLoadStatsRequest(s grpc.ClientStream, loads []*load.Data)
...
 	v2c.logger.Infof("lrs: sending LRS loads: %+v", pretty.ToJSON(req))
-	return stream.Send(req)
+	if err := stream.Send(req); err != nil {
+		return getStreamError(stream)
+	}
+	return nil
+}
+
+func getStreamError(stream lrsStream) error {
+	for {
```
Comment (easwars) [inline xds/internal/xdsclient/controller/version/v2/loadreport.go:162]: Should there be a way for this for loop to terminate other than when `Recv()` returns non-nil error?

File: xds/internal/xdsclient/controller/transport.go [hunk 4]
```diff
@@ -388,26 +388,34 @@ func (t *Controller) reportLoad(ctx context.Context, cc *grpc.ClientConn, opts c
 
 		retries++
 		lastStreamStartTime = time.Now()
-		stream, err := t.vClient.NewLoadStatsStream(ctx, cc)
+		// streamCtx is created and canceled in case we terminate the stream
+		// early for any reason, to avoid gRPC-Go leaking the RPC's monitoring
+		// goroutine.
```
Comment (easwars) [inline xds/internal/xdsclient/controller/transport.go:393]: RPC's monitoring goroutine? Which goroutine are you taking about here?

File: xds/internal/xdsclient/controller/transport.go [hunk 5]
```diff
@@ -388,26 +388,34 @@ func (t *Controller) reportLoad(ctx context.Context, cc *grpc.ClientConn, opts c
...
+		streamCtx, cancel := context.WithCancel(ctx)
+		stream, err := t.vClient.NewLoadStatsStream(streamCtx, cc)
 		if err != nil {
 			t.logger.Warningf("lrs: failed to create stream: %v", err)
+			cancel()
```
Comment (easwars) [inline xds/internal/xdsclient/controller/transport.go:398]: Should we do a `defer cancel()` and put everything that is currently inside of the for loop in an anonymous function?

File: xds/internal/xdsclient/controller/version/v2/loadreport.go (hunk 1 above)
Comment (dfawley) [reply xds/internal/xdsclient/controller/version/v2/loadreport.go:156]: The error returned from `Send` is only ever `io.EOF` in the event of the server terminating the stream or a connection loss.  This function receives the _actual_ error from the stream.

https://github.com/grpc/grpc-go/blob/64174955202ffb5ea4122e25d1aaece49cc5a3ed/stream.go#L108-L110

File: xds/internal/xdsclient/controller/transport.go (hunk 4 above)
Comment (dfawley) [reply xds/internal/xdsclient/controller/transport.go:393]: This one:

https://github.com/grpc/grpc-go/blob/64174955202ffb5ea4122e25d1aaece49cc5a3ed/stream.go#L354-L361

File: xds/internal/xdsclient/controller/version/v2/loadreport.go (hunk 2 above)
Comment (dfawley) [reply xds/internal/xdsclient/controller/version/v2/loadreport.go]: Done.

File: xds/internal/xdsclient/controller/version/v2/loadreport.go (hunk 3 above)
Comment (dfawley) [reply xds/internal/xdsclient/controller/version/v2/loadreport.go:162]: `stream.Recv` will terminate with an error if the context for the RPC is canceled, which is also how `sendLoads` exits (in `transport.go`).  So I believe this is fine.

File: xds/internal/xdsclient/controller/transport.go (hunk 5 above)
Comment (dfawley) [reply xds/internal/xdsclient/controller/transport.go]: I have mixed feelings about this sort of thing.  Done; let me know what you think.

File: xds/internal/xdsclient/controller/version/v2/loadreport.go (hunk 1 above)
Comment (easwars) [reply xds/internal/xdsclient/controller/version/v2/loadreport.go:156]: Is `Recv` always supposed to contain a more informational error on the client side, even if the error was encountered on the client? I'm thinking about the case where the error for `Send` happens on the client, but we still end up using the error we get from `Recv`. Is it possible that the former contained more contextual information?

Comment (easwars) [reply xds/internal/xdsclient/controller/version/v2/loadreport.go:156]: My question is whether we should do this only when `Send` returns an error which is no `io.EOF`.

**Pull request merged**